### PR TITLE
Add user management operations

### PR DESCRIPTION
**Pull request by FelixGail**
Adds user and permission management (#21)

I am personally unhappy with  `operationId: deleteDifferentUser`, but I don't want to break existing implementations by renaming `deleteUser`.

### CODE REVIEW

File: MusicBot.yaml [hunk 1]
```diff
@@ -795,6 +893,28 @@ components:
         refreshToken:
           description: A long-lived opaque token to refresh the access token.
           type: string
+    Permission:
+      type: string
+      description: >
+        SKIP -> Allows skipping songs;
+        DISLIKE -> Allows removing a song suggestion;
+        MOVE -> Users with this permission can alter the queue order;
+        PAUSE -> Allows pausing the playback;
```
Comment (BjoernPetersen) [inline MusicBot.yaml:902]: and resuming

File: MusicBot.yaml [hunk 2]
```diff
@@ -795,6 +893,28 @@ components:
         refreshToken:
           description: A long-lived opaque token to refresh the access token.
           type: string
+    Permission:
+      type: string
+      description: >
+        SKIP -> Allows skipping songs;
```
Comment (BjoernPetersen) [inline MusicBot.yaml:899]: From [Permission.kt](https://github.com/BjoernPetersen/MusicBot/blob/develop/src/main/kotlin/net/bjoernpetersen/musicbot/api/auth/Permission.kt) (Linked in the README.md): 

> The permission to skip the current song or remove songs from the queue.
> Note that a user is always allowed to remove a song from the queue which he added himself.

File: MusicBot.yaml [hunk 3]
```diff
@@ -795,6 +893,28 @@ components:
...
+        DISLIKE -> Allows removing a song suggestion;
+        MOVE -> Users with this permission can alter the queue order;
+        PAUSE -> Allows pausing the playback;
+        ENQUEUE -> Allows enquing songs;
+        ALTER_SUGGESTIONS -> Enquing a song while holding this permission will give feedback to suggesters
```
Comment (BjoernPetersen) [inline MusicBot.yaml:904]: "enqueuing" or "enqueueing"

File: MusicBot.yaml [hunk 4]
```diff
@@ -795,6 +893,28 @@ components:
...
+        CHANGE_VOLUME -> Allows to change the playback volume;
+        EXIT -> Allows sending a shutdown signal to the bot;
+        USER -> Allows deleting or altering other users and their permissions;
```
Comment (BjoernPetersen) [inline MusicBot.yaml:907]: This one should probably called `ADMIN`. Having the permission should NOT include the other permissions, but it effectively makes the user an admin as he is allowed to alter his own permissions.

File: MusicBot.yaml [hunk 5]
```diff
@@ -117,6 +117,104 @@ paths:
...
+      summary: Retrieves information about a user
+      description: >
+        Returns information about the user specified in the path.
+        Permission information about the user reflects the current state of the bot, permissions saved in still valid JWT tokens may vary.
+      parameters:
+        - name: username
+          in: path
+          schema:
+            type: string
+          description: The user to retrieve
```
Comment (BjoernPetersen) [inline MusicBot.yaml:132]: Note that it's case-insensitive

File: MusicBot.yaml [hunk 6]
```diff
@@ -117,6 +117,104 @@ paths:
...
+                $ref: "#/components/schemas/UserInfo"
+        401:
+          $ref: "#/components/responses/Unauthenticated"
+        403:
+          $ref: "#/components/responses/Unauthorized"
+        404:
+          description: The user does not exist
+    put:
+      operationId: setUserPermissions
+      summary: Set a users permissions
```
Comment (BjoernPetersen) [inline MusicBot.yaml:149]:
```suggestion
      summary: Set a user's permissions
```

File: MusicBot.yaml [hunk 7]
```diff
@@ -117,6 +117,104 @@ paths:
...
+      description: >
+        Updates a users permissions.
```
Comment (BjoernPetersen) [inline MusicBot.yaml:151]:
```suggestion
        Updates a user's permissions.
```

File: MusicBot.yaml [hunk 8]
```diff
@@ -117,6 +117,104 @@ paths:
...
+              schema:
+                $ref: "#/components/schemas/UserInfo"
+        401:
+          $ref: "#/components/responses/Unauthenticated"
+        403:
+          $ref: "#/components/responses/Unauthorized"
+        404:
+          description: The user does not exist
+    put:
+      operationId: setUserPermissions
```
Comment (BjoernPetersen) [inline MusicBot.yaml:148]: This operation should probably be done on `/user/{username}/permission` as it's only updating a subset of the user (his permissions).

**Review by BjoernPetersen — CHANGES_REQUESTED** [review]

File: MusicBot.yaml (hunk 4 above)
Comment (FelixGail) [reply MusicBot.yaml:907]: Do you think it is sensible to change this behaviour? For example, a user could only assign authorisations that he or she owns.

I think calling it admin could be misleading in this situation

Comment (BjoernPetersen) [reply MusicBot.yaml:908]: Only allowing a user to give permissions they own themselves seems like a good idea, I'm fine with that.

File: MusicBot.yaml [hunk 9]
```diff
@@ -117,6 +117,105 @@ paths:
...
+        404:
+          description: The user does not exist
+  "/user/{username}/permission"
+    put:
+      operationId: setUserPermissions
+      summary: Set a user's permissions
+      description: >
+        Updates a user's permissions.
+        Permissions omitted in the request will be dropped if owned by the user previously.
+        The update will reach the user, when he updates his JWT token.
```
Comment (BjoernPetersen) [inline MusicBot.yaml:154]: I think that's a very German comma

```suggestion
        The update will reach the user when he updates his JWT token.
```

File: MusicBot.yaml [hunk 10]
```diff
@@ -117,6 +117,105 @@ paths:
...
+      responses:
+        201:
+          description: The update was successful
+        401:
+          $ref: "#/components/responses/Unauthenticated"
+        403:
+          $ref: "#/components/responses/Unauthorized"
+        404:
+          description: The user does not exist
+    delete:
```
Comment (BjoernPetersen) [inline MusicBot.yaml:178]: This operation should be moved up, it's on the wrong path now (`/user/{username}/permission`)

**Review by BjoernPetersen — CHANGES_REQUESTED** [review]

File: MusicBot.yaml (hunk 4 above)
Comment (FelixGail) [reply MusicBot.yaml:908]: Should I still change `USER` to `ADMIN` or is the name acceptable, with this change?

Comment (BjoernPetersen) [reply MusicBot.yaml:908]: I wouldn't name it `ADMIN` then, but a more descriptive name would still be nice. All other permission names describe some kind of action, `USER` doesn't really convey what actions it allows.

File: MusicBot.yaml [hunk 11]
```diff
@@ -117,6 +117,105 @@ paths:
...
+              schema:
+                $ref: "#/components/schemas/UserInfo"
+        401:
+          $ref: "#/components/responses/Unauthenticated"
+        403:
+          $ref: "#/components/responses/Unauthorized"
+        404:
+          description: The user does not exist
+    delete:
+      operationId: deleteDifferentUser
```
Comment (BjoernPetersen) [inline MusicBot.yaml:148]: I'm also not that happy with this operation ID. Maybe `deleteOtherUser`?

File: MusicBot.yaml [hunk 12]
```diff
@@ -117,6 +117,105 @@ paths:
...
+          $ref: "#/components/responses/Unauthorized"
+        404:
+          description: The user does not exist
+  /users:
+    get:
+      operationId: getUsers
+      summary: Retrieve all users
+      description: >
+        Retrieves information about all full users.
+        Permission information about users reflects the current state of the bot, permissions saved in still valid jwt tokens may vary.
```
Comment (BjoernPetersen) [inline MusicBot.yaml:205]:
```suggestion
        Permission information about users reflects the current state of the bot, permissions saved in still valid JWT tokens may vary.
```

File: MusicBot.yaml [hunk 13]
```diff
@@ -795,6 +894,28 @@ components:
...
+      enum:
+        - SKIP
+        - DISLIKE
+        - MOVE
+        - PAUSE
+        - ENQUEUE
+        - ALTER_SUGGESTIONS
+        - CHANGE_VOLUME
+        - EXIT
+        - USER
```
Comment (BjoernPetersen) [inline MusicBot.yaml:918]: These should all be lower case as that's the value actually transmitted via the API (see `label` in [Permissions.kt](https://github.com/BjoernPetersen/MusicBot/blob/develop/src/main/kotlin/net/bjoernpetersen/musicbot/api/auth/Permission.kt))

**Review by BjoernPetersen — CHANGES_REQUESTED** [review]

File: MusicBot.yaml (hunk 4 above)
Comment (FelixGail) [reply MusicBot.yaml:908]: `MANAGE_USERS`?

Comment (BjoernPetersen) [reply MusicBot.yaml:908]: Yeah, that's nice enough.

**Review by BjoernPetersen — CHANGES_REQUESTED** [review]
The operation descriptions should mention the required permission ("manage_users").

Do you think it might be problematic to effectively allow users to impersonate another user by deleting and recreating them? Maybe that should be a separate permission.

How do we handle attempts to change permissions of a user with more permission than the managing user? E.g.:

```
userA.permissions = [manage_users, skip, pause]
userB.permissions = [skip, pause, dislike, move, enqueue, alter_suggestions]
userA.setPermissions(userB, [skip, pause])
// What are userB's permissions now? Just skip and pause?
```

**Comment by FelixGail** [comment]
> Do you think it might be problematic to effectively allow users to impersonate another user by deleting and recreating them? Maybe that should be a separate permission.

This is indeed problematic. I think a separate permission for user deletions is the most straightforward change.

> How do we handle attempts to change permissions of a user with more permission than the managing user? E.g.:
> ```
> userA.permissions = [manage_users, skip, pause]
> userB.permissions = [skip, pause, dislike, move, enqueue, alter_suggestions]
> userA.setPermissions(userB, [skip, pause])
> // What are userB's permissions now? Just skip and pause?
> ```

Do you think this could be solved by allowing a user to only **change** those permissions he owns? In the above example `userA` would only be able to bestow the `skip` and `pause` onto `userB`, but not remove any of the permissions he doesn't own.

I intentionally skipped  the `manage_users` permission in the above example, because if we handle permissions this way this permission needs to be separately. Otherwise we would create a new loophole, where `userA` gives `userB` the `manage_users` permission who in turn can bestow all of his permissions.

I can think of three possible solutions:

- _The Teamspeak way:_ We could add a new admin permission
- We change the `manage_users` permission to an `admin` permission, giving the user all other permissions implicitly
- In order to bestow the `manage_users` permission a user needs all other pemissions

**Comment by BjoernPetersen** [comment]
> Otherwise we would create a new loophole, where userA gives userB the manage_users permission who in turn can bestow all of his permissions.

I don't think that's a loophole, really. Users who are given the `manage_users` permission may give that permissions to other users, but they could only escalate their own permissions by that if the other user with more permissions colludes with them. I'd personally be fine with that risk.

> Do you think this could be solved by allowing a user to only change those permissions he owns? In the above example userA would only be able to bestow the skip and pause onto userB, but not remove any of the permissions he doesn't own.

That's a good approach, but I don't think it's enough. That limitation would still allow a user with `manage_users` permissions to remove certain permissions of any user, even the server owner who gave himself every possible permission.

It's a puzzler, really. This is going to need some time to think about it.

**Comment by BjoernPetersen** [comment]
After some discussion with a flatmate of mine, I think the best solution is making the `manage_users` permission an `admin` permission. The permission allows a user to grant or revoke any permissions for himself or other users.

When a user is granted the `admin` permission he is automatically granted all other permissions. If more possible permissions are added to the system later on, they'll have to be granted manually.

While the `admin` permission allows the user to escalate their own permissions, it doesn't in itself include any other permission, i.e. a user with the `admin` permission but without the `enqueue` permission will not be able to enqueue songs.

To prevent "coups", I suggest the following limitation:
An admin can not edit the permissions of another admin, except for his own. Removing admins will thus not be possible via the API, although admins may step down of course.

**Comment by FelixGail** [comment]
FYI: The editorconfig check runs through, but the check is running on the fork somehow: https://github.com/FelixGail/MusicBot-API/actions/runs/363214297

**Comment by BjoernPetersen** [comment]
Sorry about that. Updating your branch should fix the problem.

**Comment by BjoernPetersen** [comment]
Okay, I really want to merge this branch. I do have one more idea though. It's kind of a rehash of some of our ideas to the current solution:

## Introduce a `manage_admins` permission

Introducing this permission would allow almost all user management to be done via the API.

### Granted rights

- Grant and revoke the `admin` permission
- Grant, but **not revoke** the `manage_admins` permission 
  - Analogously to the previous approach for admins
  - The revocation aspect limits the API a bit, but prevents "coups". I'm not sure about it yet, so input is welcome.

### Changes to `admin`

- Users with the `admin` permission can not add new admins anymore
- Making a user admin doesn't grant him the `manage_admins` permission

### Benefits

- Appointing an admin temporarily is much more viable
  - Removing the `admin` permission of a user doesn't leave him a time window in which he could re-add the permission with the existing JWT
- "Rogue" admins can't do as much damage
- The `manage_admins` permission can be reserved for especially trusted users (maybe the server owner even completely reserves it for himself)
  - This is what makes me doubt the `manage_admin` revocation limitation

**Comment by BjoernPetersen** [comment]
Additionally, we should add a limitation to the `deleteOtherUser` operation: Only non-admins may be deleted.

**Comment by FelixGail** [comment]
> Introduce a manage_admins permission

That's a nice solution

>     * The `manage_admins` permission can be reserved for especially trusted users (maybe the server owner even completely reserves it for himself)
>       
>       * This is what makes me doubt the `manage_admin` revocation limitation

I don't think this restriction is needed. Not having it could even proof beneficial as unused accounts could be removed over the API.
To match the intention of this permission it could be named `OWNER`, setting as guideline that all users with this permission should be able to access the bot directly anyway.

**Comment by BjoernPetersen** [comment]
Yeah, I also toyed with `superuser`, but `owner` is probably most obvious.

File: MusicBot.yaml [hunk 14]
```diff
@@ -117,6 +117,113 @@ paths:
...
+        OWNERs may delete all other users.
+      parameters:
+        - name: username
+          in: path
+          schema:
+            type: string
+          description: The user to delete
+          required: true
+      responses:
+        201:
```
Comment (BjoernPetersen) [inline MusicBot.yaml:164]:
```suggestion
        204:
```

`201` is "Created"

File: MusicBot.yaml [hunk 15]
```diff
@@ -117,6 +117,113 @@ paths:
...
+          required: true
+      requestBody:
+        content:
+          application/json:
+            schema:
+              type: array
+              items:
+                $ref: '#/components/schemas/Permission'
+      responses:
+        201:
```
Comment (BjoernPetersen) [inline MusicBot.yaml:198]:
```suggestion
        204:
```

File: MusicBot.yaml [hunk 16]
```diff
@@ -809,6 +916,33 @@ components:
...
+          Note that a user is always allowed to remove a song from the queue which he added himself.;
+        DISLIKE -> The permission to remove songs from the upcoming suggestions of a suggester;
+        MOVE -> Users with this permission can alter the queue order;
+        PAUSE -> Allows pausing/resuming the playback;
+        ENQUEUE -> Allows enqueueing songs;
+        ALTER_SUGGESTIONS -> Enquing a song while holding this permission will give feedback to suggesters
+        CHANGE_VOLUME -> Allows to change the playback volume;
+        EXIT -> Allows sending a shutdown signal to the bot;
+        ADMIN -> Allows deleting or altering other users and their permissions.
+          Granting a user the admin permission will automatically grant him all other current permissions.
```
Comment (BjoernPetersen) [inline MusicBot.yaml:932]:
```suggestion
          Granting a user the admin permission will automatically grant him all other current permissions except OWNER.
```

File: MusicBot.yaml [hunk 17]
```diff
@@ -809,6 +916,33 @@ components:
...
+          Users holding the admin permission can not alter the permissions of other admins;
```
Comment (BjoernPetersen) [inline MusicBot.yaml:933]:
```suggestion
          Users holding the admin permission can not alter the permissions of or delete other admins;
```

**Review by BjoernPetersen — CHANGES_REQUESTED** [review]

**Comment by BjoernPetersen** [comment]
The docs should also mention that admins (not owners) can't give others the ADMIN permission.

File: MusicBot.yaml [hunk 18]
```diff
@@ -809,6 +916,33 @@ components:
...
+        MOVE -> Users with this permission can alter the queue order;
+        PAUSE -> Allows pausing/resuming the playback;
+        ENQUEUE -> Allows enqueueing songs;
+        ALTER_SUGGESTIONS -> Enquing a song while holding this permission will give feedback to suggesters
+        CHANGE_VOLUME -> Allows to change the playback volume;
+        EXIT -> Allows sending a shutdown signal to the bot;
+        ADMIN -> Allows deleting or altering other users (except OWNERs) and their permissions.
+          Granting a user the admin permission will automatically grant him all other current permissions except OWNER.
+          Users holding the admin permission can not alter the permissions of or delete other admins;
+        OWNER -> Allows altering permissions or deleting users holding the ADMIN or OWNER permission.
```
Comment (BjoernPetersen) [inline MusicBot.yaml:934]:
```suggestion
        OWNER -> Allows granting the ADMIN permission, altering permissions of admins and owners or deleting ADMIN or OWNER users.
```

**Review by BjoernPetersen — APPROVED** [review]

File: MusicBot.yaml [hunk 19]
```diff
@@ -809,6 +916,33 @@ components:
...
+        MOVE -> Users with this permission can alter the queue order;
+        PAUSE -> Allows pausing/resuming the playback;
+        ENQUEUE -> Allows enqueueing songs;
+        ALTER_SUGGESTIONS -> Enquing a song while holding this permission will give feedback to suggesters
+        CHANGE_VOLUME -> Allows to change the playback volume;
+        EXIT -> Allows sending a shutdown signal to the bot;
+        ADMIN -> Allows deleting or altering other users (except OWNERs) and their permissions.
+          Granting a user the admin permission will automatically grant him all other current permissions except OWNER.
+          Users holding the admin permission can not alter the permissions of or delete other admins;
+        OWNER -> Allows granting the ADMIN permission, altering permissions of admins and owners or deleting ADMIN or OWNER users.
```
Comment (BjoernPetersen) [inline MusicBot.yaml:934]:
```suggestion
        OWNER -> Allows granting the ADMIN permission, altering permissions of admins and owners or
          deleting ADMIN or OWNER users.
```

**Comment by BjoernPetersen** [comment]
Thanks for your contribution!

**Pull request merged**